### ISSUE TITLE
What's purpose to calculate fp32_grads from fp16_grads on post_backward_with_master_weights()

### ISSUE THREAD

**Issue by un-knight** [NVIDIA/apex#507]
Under the `O2` configuration, I found that there is a `post_backward_master_weight()` which calculates `new_fp32_grads` for fp32_param from `all_fp32_from_fp16_params` by unscaling fp16 gradients, but thoes fp32_params seem don't be used in `optimizer.step()` stage. So what is the purpose to calculate fp32_grads for fp32_param except for overflow detection?

**Comment by FDecaYed** [comment]
I assume you are talking about this?
https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L180
You're correct, this is unscaling fp16 grad and casting to fp32. But it is actually used to update its corresponding weight(master weight), since weight optimizer sees has already been modified to this in `lazy_init_with_master_weights` here
https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L28
As a result, these 3 lines in `new_step`
https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L349
 will trigger the update on master weight and cast it into model weight(which optimizer no longer sees)

**Comment by un-knight** [comment]
@FDecaYed Thanks for your reply! This method https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L351 will copy and cast fp32 master weight into fp16 model weight, and then the code **reset master weight grads to None** in [following lines](https://github.com/NVIDIA/apex/blob/master/apex/amp/_process_optimizer.py#L354), and finally, the native optimizer.step() be called and executed which will perform parameters update with fp16 scaling grads and fp16 model weight. So I don't know how master weight grads are used to update master weights?

**Comment by FDecaYed** [comment]
line 349 `retval = old_step()` is the line that uses fp32 grad to update fp32 weight. It is just native optimizer.step(), which operates on master weight/grad.
then line 351 copy the updated fp32 weight(what optimizer sees) into fp16 weight(what model sees)
next, it set fp32 grad to None to save memory, since it is no longer needed now(arguably this is a bug? @mcarilli ). But it should not affect common use case where user zero grad after when step() is called.

**Comment by un-knight** [comment]
Oh I see! Thanks!

**Comment by abhiagwl4262** [comment]
@FDecaYed @un-knight Please confirm that what I am inferring from this is correct or not -  
1. During training fp32 weights are updated and low precision fp16 weights are generated by type casting of updated fp32 to fp16 ?
2. fp16 weights are not updated using backpropagation? 

Am I correct ?

**Comment by FDecaYed** [comment]
@abhiagwl4262 your understanding is correct. That is what happens when master weight is used. We recommend master weight most of the time since it is relatively low cost to perf and avoid lots of trouble.
For more info see section 3.1
https://arxiv.org/pdf/1710.03740.pdf
